src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValuesFactory.java

```diff
@@ -72,6 +74,37 @@ public static <T> KNNVectorValues<T> getVectorValues(final FieldInfo fieldInfo,
         return getVectorValues(FieldInfoExtractor.extractVectorDataType(fieldInfo), vectorValuesIterator);
     }
 
+    /**
+     * Returns a {@link KNNVectorValues} for the given {@link FieldInfo} and {@link LeafReader}
+     *
+     * @param fieldInfo {@link FieldInfo}
+     * @param docValuesProducer {@link DocValuesProducer}
+     * @param knnVectorsReader {@link KnnVectorsReader}
+     * @return {@link KNNVectorValues}
+     */
+    public static <T> KNNVectorValues<T> getVectorValues(
+        final FieldInfo fieldInfo,
+        final DocValuesProducer docValuesProducer,
+        final KnnVectorsReader knnVectorsReader
+    ) throws IOException {
+        final DocIdSetIterator docIdSetIterator;
+        if (fieldInfo.hasVectorValues() && knnVectorsReader != null) {
+            if (fieldInfo.getVectorEncoding() == VectorEncoding.BYTE) {
+                docIdSetIterator = knnVectorsReader.getByteVectorValues(fieldInfo.getName());
+            } else if (fieldInfo.getVectorEncoding() == VectorEncoding.FLOAT32) {
+                docIdSetIterator = knnVectorsReader.getFloatVectorValues(fieldInfo.getName());
+            } else {
+                throw new IllegalArgumentException("Invalid Vector encoding provided, hence cannot return VectorValues");
+            }
+        } else if (docValuesProducer != null) {
+            docIdSetIterator = docValuesProducer.getBinary(fieldInfo);
+        } else {
+            throw new IllegalArgumentException("Field does not have vector values and DocValues");
```

Comment:
nit: Field should have either vector values or DocValues

Reply:
Will update in another PR